scalar_udf.go

```diff
@@ -0,0 +1,247 @@
+package duckdb
+
+// Related issues: https://golang.org/issue/19835, https://golang.org/issue/19837.
+
+/*
+#include <duckdb.h>
+
+void scalar_udf_callback(duckdb_function_info, duckdb_data_chunk, duckdb_vector);
+void scalar_udf_delete_callback(void *);
+
+typedef void (*scalar_udf_callback_t)(duckdb_function_info, duckdb_data_chunk, duckdb_vector);
+*/
+import "C"
+
+import (
+	"database/sql"
+	"database/sql/driver"
+	"runtime/cgo"
+	"unsafe"
+)
+
+type rowFn func(args []driver.Value) (any, error)
+
+type ScalarFuncConfig struct {
+	InputTypeInfos []TypeInfo
+	ResultTypeInfo TypeInfo
+
+	VariadicTypeInfo    *TypeInfo
+	Volatile            bool
+	SpecialNullHandling bool
+}
+
+type ScalarFuncExecutor struct {
+	RowExecutor rowFn
+}
+
+type ScalarFunc interface {
+	Config() ScalarFuncConfig
+	Executor() ScalarFuncExecutor
+}
+
+func setFuncError(function_info C.duckdb_function_info, msg string) {
+	err := C.CString(msg)
+	C.duckdb_scalar_function_set_error(function_info, err)
+	C.duckdb_free(unsafe.Pointer(err))
+}
+
+//export scalar_udf_callback
+func scalar_udf_callback(function_info C.duckdb_function_info, input C.duckdb_data_chunk, output C.duckdb_vector) {
+	extraInfo := C.duckdb_scalar_function_get_extra_info(function_info)
+
+	// extraInfo is a void* pointer to our ScalarFunc.
+	h := *(*cgo.Handle)(unsafe.Pointer(extraInfo))
+	function := h.Value().(ScalarFunc)
+
+	// Initialize the input chunk.
+	var inputChunk DataChunk
+	if err := inputChunk.initFromDuckDataChunk(input, false); err != nil {
+		setFuncError(function_info, getError(errAPI, err).Error())
+		return
+	}
+
+	// Initialize the output chunk.
+	var outputChunk DataChunk
+	if err := outputChunk.initFromDuckVector(output, true); err != nil {
+		setFuncError(function_info, getError(errAPI, err).Error())
+		return
+	}
+
+	// Execute the user-defined scalar function for each row.
+	executor := function.Executor()
+	values := make([]driver.Value, len(inputChunk.columns))
+	columnCount := len(values)
+
+	var err error
+	for rowIdx := 0; rowIdx < inputChunk.GetSize(); rowIdx++ {
```

Comment:
I think this currently calls cgo, if you put `inputChunk.GetSize()` outside of the loop it should improve performance

Reply:
Oh, great catch!